extensions/libarchive/CompressContent.h

```diff
@@ -341,6 +346,73 @@ class CompressContent: public core::Processor {
     }
   };
 
+  class GzipWriteCallback : public OutputStreamCallback {
+   public:
+    GzipWriteCallback(std::string compress_mode, int64_t compress_level, std::shared_ptr<core::FlowFile> flow, std::shared_ptr<core::ProcessSession> session)
+      : logger_(logging::LoggerFactory<CompressContent>::getLogger())
+      , compress_mode_(std::move(compress_mode))
+      , compress_level_(compress_level)
+      , flow_(std::move(flow))
+      , session_(std::move(session))
+    {
+    }
+
+    ~GzipWriteCallback() override = default;
+
+    std::shared_ptr<logging::Logger> logger_;
+    std::string compress_mode_;
+    int64_t compress_level_;
+    std::shared_ptr<core::FlowFile> flow_;
+    std::shared_ptr<core::ProcessSession> session_;
+    bool success_{false};
+
+    int64_t process(std::shared_ptr<io::BaseStream> outputStream) override {
+      class ReadCallback : public InputStreamCallback {
+       public:
+        ReadCallback(GzipWriteCallback& writer, std::shared_ptr<io::BaseStream> outputStream)
+          : writer_(writer)
+          , outputStream_(std::move(outputStream)) {
+        }
+
+        int64_t process(std::shared_ptr<io::BaseStream> inputStream) override {
+          std::vector<uint8_t> buffer(16 * 1024U);
+          int64_t read_size = 0;
+          while (read_size < writer_.flow_->getSize()) {
+            int ret = inputStream->read(buffer.data(), buffer.size());
+            if (ret < 0) {
+              return -1;
+            } else if (ret == 0) {
+              break;
+            } else {
+              if (outputStream_->writeData(buffer.data(), ret) != ret) {
+                return -1;
+              }
+              read_size += ret;
+            }
+          }
+          outputStream_->closeStream();
```

Comment:
The generic answer is that our Stream API is not very highly (or properly) defined and whatever `closeStream` does, or when it should be called is implementation-defined.
In this particular case we have to call `closeStream` in the successful case, because that's how we signal the end of the input data and force zlib to flush its buffers and write out the end of the compressed stream.